pkg/common/commands/resetoffset/controller/offsets.go

```diff
@@ -0,0 +1,230 @@
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"knative.dev/pkg/logging"
+
+	kafkav1alpha1 "knative.dev/eventing-kafka/pkg/apis/kafka/v1alpha1"
+)
+
+// SaramaNewClientFnType defines the Sarama NewClient() function signature.
+type SaramaNewClientFnType func([]string, *sarama.Config) (sarama.Client, error)
+
+// SaramaNewClientFn is a reference to the Sarama NewClient() function used
+// when reconciling offsets which facilitates stubbing in unit tests.
+var SaramaNewClientFn SaramaNewClientFnType = sarama.NewClient
+
+// SaramaNewOffsetManagerFromClientFnType defines the Sarama NewOffsetManagerFromClient() function signature.
+type SaramaNewOffsetManagerFromClientFnType func(group string, client sarama.Client) (sarama.OffsetManager, error)
+
+// SaramaNewOffsetManagerFromClientFn is a reference to the Sarama NewOffsetManagerFromClient()
+// function used when reconciling offsets which facilitates stubbing in unit tests.
+var SaramaNewOffsetManagerFromClientFn SaramaNewOffsetManagerFromClientFnType = sarama.NewOffsetManagerFromClient
+
+// updateOffsetsError is the general error for any failure in updating offsets, exposed for test use.
+var updateOffsetsError = fmt.Errorf("failed to update Offsets")
+
+// reconcileOffsets updates the Offsets of all Partitions for the specified
+// Topic / ConsumerGroup to the Offset value corresponding to the specified
+// offsetTime (millis since epoch) and return OffsetMappings of the old/new
+// state.  An error will be returned and the Offsets will not be committed
+// if any problems occur.
+func (r *Reconciler) reconcileOffsets(ctx context.Context, topicName string, groupId string, offsetTime int64) ([]kafkav1alpha1.OffsetMapping, error) {
+
+	// Get The Logger From The Context & Enhance The With Parameters
+	logger := logging.FromContext(ctx).Desugar().With(
+		zap.String("Topic", topicName),
+		zap.String("Group", groupId),
+		zap.Int64("Time", offsetTime))
+
+	// Initialize A New Sarama Client
+	//
+	// ResetOffset is an infrequently used feature so there is no need for
+	// reuse, and there are also "broken-pipe" failures (non-recoverable)
+	// after periods of inactivity to deal with...
+	//   https://github.com/Shopify/sarama/issues/1162
+	//   https://github.com/Shopify/sarama/issues/866
+	saramaClient, err := SaramaNewClientFn(r.kafkaBrokers, r.saramaConfig)
+	defer safeCloseSaramaClient(logger, saramaClient)
+	if saramaClient == nil || err != nil {
+		logger.Error("Failed to create a new Sarama Client", zap.Error(err))
+		return nil, err
+	}
+
+	// Get The Partitions Of The Specified Kafka Topic
+	partitions, err := saramaClient.Partitions(topicName)
+	if err != nil {
+		logger.Error("Failed to determine Partitions for Topic", zap.Error(err))
+		return nil, err
+	}
+	logger.Info("Found Topic Partitions", zap.Any("Partitions", partitions))
```

Comment:
debug?

Reply:
Sure I can change it.  I tend towards noisy logs in Controllers as they are Control-Plane.  This one in particular will be used infrequently (downtime recovery and other edge cases).